File: POSmaisgestao/app/src/main/java/com/pos_mais_gestao/hardware/balanca/protocol/Prt3Parser.java
package com.pos_mais_gestao.hardware.balanca.protocol;

import com.pos_mais_gestao.domain.balanca.BalancaEstado;
import com.pos_mais_gestao.domain.balanca.BalancaLeitura;
import java.io.ByteArrayOutputStream;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/** Parser incremental do Prt3: STX + cinco caracteres ASCII + ETX. */
public final class Prt3Parser {
    private static final int PAYLOAD_SIZE = 5;
    private static final int MAX_FRAME_SIZE = 32;
    private final ByteArrayOutputStream frame = new ByteArrayOutputStream();
    private boolean recebendo;

    public synchronized List<BalancaLeitura> aceitar(byte[] dados, int tamanho) {
        List<BalancaLeitura> leituras = new ArrayList<>();
        if (dados == null || tamanho <= 0) {
            return leituras;
        }
        int limite = Math.min(tamanho, dados.length);
        for (int i = 0; i < limite; i++) {
            byte atual = dados[i];
            if (atual == Prt3Protocol.STX) {
                frame.reset();
                recebendo = true;
                continue;
            }
            if (!recebendo) {
                continue;
            }
            if (atual == Prt3Protocol.ETX) {
                leituras.add(parsearFrame(frame.toByteArray()));
                frame.reset();
                recebendo = false;
                continue;
            }
            if (frame.size() >= MAX_FRAME_SIZE) {
                frame.reset();
                recebendo = false;
                leituras.add(BalancaLeitura.estado(BalancaEstado.ERRO, "Frame Prt3 excedeu o limite"));
                continue;
            }
            frame.write(atual);
        }
        return leituras;
    }

    public synchronized void resetar() {
        frame.reset();
        recebendo = false;
    }

    public static BalancaLeitura parsearFrame(byte[] payload) {
        if (payload == null || payload.length != PAYLOAD_SIZE) {
            return BalancaLeitura.estado(BalancaEstado.ERRO, "Frame Prt3 incompleto");
        }
        String valor = new String(payload, StandardCharsets.US_ASCII);
        if ("IIIII".equals(valor)) {
            return BalancaLeitura.estado(BalancaEstado.INSTAVEL, "Peso instável");
        }
        if ("NNNNN".equals(valor)) {
            return BalancaLeitura.estado(BalancaEstado.PESO_NEGATIVO, "Peso negativo");
        }
        if ("SSSSS".equals(valor)) {
            return BalancaLeitura.estado(BalancaEstado.SOBRECARGA, "Sobrecarga da balança");
        }
        if (!valor.matches("\\d{5}")) {
            return BalancaLeitura.estado(BalancaEstado.ERRO, "Caracteres inválidos no frame Prt3");
        }
        BigDecimal kg = new BigDecimal(valor).movePointLeft(3).setScale(3, RoundingMode.UNNECESSARY);
        return BalancaLeitura.peso(kg);
    }
}

File: estacao-balanca/lib/features/balanca/prt3_parser.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:estacao_balanca/features/balanca/balanca_leitura.dart';
import 'package:estacao_balanca/features/balanca/prt3_protocol.dart';

/// Parser incremental Prt3: STX + cinco caracteres ASCII + ETX.
class Prt3Parser {
  static const int payloadSize = 5;
  static const int maxFrameSize = 32;

  final BytesBuilder _frame = BytesBuilder(copy: false);
  bool _recebendo = false;

  List<BalancaLeitura> aceitar(List<int> dados) {
    final leituras = <BalancaLeitura>[];
    for (final atual in dados) {
      if (atual == Prt3Protocol.stx) {
        _frame.clear();
        _recebendo = true;
        continue;
      }
      if (!_recebendo) continue;
      if (atual == Prt3Protocol.etx) {
        leituras.add(parsearFrame(_frame.takeBytes()));
        _recebendo = false;
        continue;
      }
      if (_frame.length >= maxFrameSize) {
        _frame.clear();
        _recebendo = false;
        leituras.add(
          BalancaLeitura.estado(
            BalancaEstado.erro,
            'Frame Prt3 excedeu o limite',
          ),
        );
        continue;
      }
      _frame.addByte(atual);
    }
    return leituras;
  }

  void resetar() {
    _frame.clear();
    _recebendo = false;
  }

  static BalancaLeitura parsearFrame(List<int> payload) {
    if (payload.length != payloadSize) {
      return BalancaLeitura.estado(
        BalancaEstado.erro,
        'Frame Prt3 incompleto',
      );
    }
    final valor = ascii.decode(payload, allowInvalid: true);
    if (valor == 'IIIII') {
      return BalancaLeitura.estado(BalancaEstado.instavel, 'Peso instável');
    }
    if (valor == 'NNNNN') {
      return BalancaLeitura.estado(
        BalancaEstado.pesoNegativo,
        'Peso negativo',
      );
    }
    if (valor == 'SSSSS') {
      return BalancaLeitura.estado(
        BalancaEstado.sobrecarga,
        'Sobrecarga da balança',
      );
    }
    if (!RegExp(r'^\d{5}$').hasMatch(valor)) {
      return BalancaLeitura.estado(
        BalancaEstado.erro,
        'Caracteres inválidos no frame Prt3',
      );
    }
    final kg = int.parse(valor) / 1000.0;
    return BalancaLeitura.peso(double.parse(kg.toStringAsFixed(3)));
  }
}
